tests/integration_tests/performance/test_net_tput.py

```diff
@@ -0,0 +1,290 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Tests the network throughput overhead added by Firecracker."""
+
+import json
+import time
+import statistics
+from subprocess import run
+import pytest
+
+import host_tools.network as net_tools
+
+RUNS = 1
+
+IPERF = 'iperf3'
+ITERATION_DURATION = 1
+WINDOW_SIZE = 10
+
+# BW in Mbits.
+BANDWIDTHS = {
+    'udp': [100, 5000, 10000],
+    'tcp': [5000, 10000, 20000, 0]
+}
+
+# Packet size used for TCP.
+PACKET_SIZE = [128, 256, 1024, 1500]
+
+# From server to client and reversed.
+MODES = ['h2g', 'g2h']
+# Corresponding MODES for iperf options.
+MODES_OPTIONS = {
+    'h2g': '-R',
+    'g2h': ''
+}
+
+PROTOCOL = ['udp', 'tcp']
+PROTOCOL_OPTIONS = {
+    'tcp': '',
+    'udp': '-u'
+}
+
+# Skip the first OMIT_TIME seconds when testing TCP
+# to avoid slow start.
+OMIT_TIME = 0
+
+PINNING_CPU = 1
+
+OFFSET_PORT = 23412
+
+OUTPUT_JSON = 'test_net_tput_{}_iperfs.json'
+
+
+class Statistics:
+    """Holds results from multiple runs of the same test."""
+
+    def __init__(self):
+        """Initialize variables used to compute statistics."""
+        self.values = []
+        self.count = 0
+        self.dev = 0
+        self.mean = self.min = self.max = self.median = 0
+
+    def add(self, value):
+        """Add an observed value to the list."""
+        self.values.append(value)
+        self.count += 1
+
+    def compute_stats(self):
+        """Compute the statistics on the given data."""
+        self.mean = statistics.mean(self.values)
+        self.median = statistics.median(self.values)
+        self.dev = statistics.stdev(self.values)
+        self.min = min(self.values)
+        self.max = max(self.values)
+
+
+def create_tcp_statistics():
+    """Create a map which holds metric statistics registered for TCP."""
+    return {
+        'tput': Statistics(),
+        'retransmits': Statistics(),
+    }
+
+
+def create_udp_statistics():
+    """Create a map which holds metric statistics registered for UDP."""
+    return {
+        'tput': Statistics(),
+        'lost': Statistics(),
+        'pps': Statistics(),
+        'jitter': Statistics(),
+    }
+
+
+def spawn_host_iperf(netns_cmd_prefix):
+    """Start iperf in server mode after killing any leftover iperf daemon."""
+    iperf_cmd = 'pkill {}'.format(IPERF)
+
+    # Don't check the result of this command because it can fail if no iperf
+    # is running.
+    run(iperf_cmd, shell=True, check=False)
+
+    iperf_cmd = '{} {} -sD'.format(netns_cmd_prefix, IPERF)
+    run(iperf_cmd, shell=True, check=True)
+
+    # Wait for the iperf daemon to start.
+    time.sleep(2)
```

Comment:
Can we do something more deterministic here ? Does the `run` method return any result that would help us identify the moment when the iper process started ?

Reply:
My guess would be that the server is spawned as another process and then the parent returns its exit code and goes away, so I think we can't find out with `run`'s help, but I will check it out closely.

Reply:
You can look busy wait on netstat -lp to see if there a listen socket opened for the server port. If there is a socket opened than you are good to go. This is the correct method to solve this.

Reply:
I tried this too, but the command has no entries when running it inside the docker image, so I couldn't see when the server started.

Reply:
then try doing a connect to server in a loop and then close it